libDF/src/tract.rs

```diff
@@ -0,0 +1,724 @@
+                    self.nb_erb,
+                    self.min_nb_erb_freqs,
+                );
+                state.init_norm_states(self.nb_df);
+                self.df_states.push(state)
+            }
+        }
+        self.spec_buf = Tensor::zero::<f32>(&spec_shape)?;
+        self.erb_buf = Tensor::zero::<f32>(&[ch, 1, 1, self.nb_erb])?;
+        self.cplx_buf = Tensor::zero::<f32>(&[ch, 1, self.nb_df, 2])?;
+
+        // for _ in 0..(enc_delay.max(erb_dec_delay.max(df_dec_delay))) {
+        //     let mut enc_emb = self.enc.run(tvec!(
+        //         self.erb_buf.clone(),
+        //         self.cplx_buf.clone().permute_axes(&[0, 3, 1, 2])?
+        //     ))?;
+        //     dbg!(enc_emb.pop());
+        //     let c0 = enc_emb.pop().unwrap().into_tensor();
+        //     let cemb = enc_emb.pop().unwrap().into_tensor();
+        //     let e3 = enc_emb.pop().unwrap().into_tensor();
+        //     let e2 = enc_emb.pop().unwrap().into_tensor();
+        //     let e1 = enc_emb.pop().unwrap().into_tensor();
+        //     let e0 = enc_emb.pop().unwrap().into_tensor();
+
+        //     self.erb_dec.run(tvec!(cemb.clone(), e3, e2, e1, e0))?;
+
+        //     self.df_dec.run(tvec!(cemb.clone(), c0))?;
+        // }
+
+        Ok(())
+    }
+
+    pub fn process(&mut self, noisy: ArrayView2<f32>, mut enh: ArrayViewMut2<f32>) -> Result<()> {
+        #[cfg(feature = "timings")]
+        let t0 = Instant::now();
+        let ch = noisy.len_of(Axis(0));
+        debug_assert_eq!(noisy.len_of(Axis(0)), enh.len_of(Axis(0)));
+        debug_assert_eq!(noisy.len_of(Axis(1)), enh.len_of(Axis(1)));
+        debug_assert_eq!(noisy.len_of(Axis(1)), self.hop_size);
+
+        self.rolling_spec_buf.rotate_left(1);
+        for (ns_ch, mut rbuf, mut erb_ch, mut cplx_ch, state) in zip5(
+            noisy.axis_iter(Axis(0)),
+            self.rolling_spec_buf
+                .back_mut()
+                .unwrap()
+                .to_array_view_mut()?
+                .axis_iter_mut(Axis(0)),
+            self.erb_buf.to_array_view_mut()?.axis_iter_mut(Axis(0)),
+            self.cplx_buf.to_array_view_mut()?.axis_iter_mut(Axis(0)),
+            self.df_states.iter_mut(),
+        ) {
+            let spec = as_slice_mut_complex(rbuf.as_slice_mut().unwrap());
+            state.analysis(ns_ch.as_slice().unwrap(), spec);
+            state.feat_erb(spec, self.alpha, erb_ch.as_slice_mut().unwrap());
+            // TODO: Workaround transpose by directly computing transposed complex features:
+            // state.feat_cplx_t(&spec[..nb_df], alpha, cplx_buf_t.as_slice_mut()?);
+            state.feat_cplx(
+                &spec[..self.nb_df],
+                self.alpha,
+                as_slice_mut_complex(cplx_ch.as_slice_mut().unwrap()),
+            );
+        }
+
+        #[cfg(feature = "timings")]
+        let t1 = Instant::now();
+        // Run encoder
+        let mut enc_emb = self.enc.run(tvec!(
```

Comment:
Compute the encoder embedding. This model uses tract-pulse for its 3x3 convolution in the first layer.